source/common/router/header_formatter.cc

```diff
@@ -2,13 +2,109 @@
 
 #include <string>
 
+#include "envoy/common/optional.h"
+
 #include "common/access_log/access_log_formatter.h"
+#include "common/common/utility.h"
+#include "common/config/metadata.h"
+#include "common/json/json_loader.h"
 
 #include "fmt/format.h"
 
 namespace Envoy {
 namespace Router {
 
+namespace {
+
+std::string formatUpstreamMetadataParseException(const std::string& params,
+                                                 const EnvoyException* cause = nullptr) {
+  std::string reason;
+  if (cause != nullptr) {
+    reason = fmt::format(", because {}", cause->what());
+  }
+
+  return fmt::format("Incorrect header configuration. Expected format "
+                     "UPSTREAM_METADATA([\"namespace\", \"k\", ...]), actual format "
+                     "UPSTREAM_METADATA{}{}",
+                     params, reason);
+}
+
+std::function<std::string(const Envoy::AccessLog::RequestInfo&)>
+parseUpstreamMetadataField(const std::string& params_str) {
```

Comment:
Can you add a comment with an example string of the correct format near the top of the function, so I can quickly see visually what it is this function is parsing?